tests/lxd/test_lxd_image_vault.cpp

```diff
@@ -0,0 +1,392 @@
+
+    EXPECT_EQ(image.id, mpt::default_id);
+    EXPECT_EQ(image.stream_location, mpt::default_stream_location);
+    EXPECT_EQ(image.original_release, "18.04 LTS");
+    EXPECT_EQ(image.release_date, mpt::default_version);
+}
+
+TEST_F(LXDImageVault, throws_with_invalid_alias)
+{
+    ON_CALL(host, info_for(_)).WillByDefault([this](auto query) -> mp::optional<mp::VMImageInfo> {
+        if (query.release != "bionic")
+        {
+            return mp::nullopt;
+        }
+
+        return host.mock_image_info;
+    });
+
+    ON_CALL(*mock_network_access_manager.get(), createRequest(_, _, _)).WillByDefault([](auto...) {
+        return new mpt::MockLocalSocketReply(mpt::not_found_data, QNetworkReply::ContentNotFoundError);
+    });
+
+    mp::Query query{"", "xenial", false, "release", mp::Query::Type::Alias};
+
+    mp::LXDVMImageVault image_vault{hosts, mock_network_access_manager.get(), base_url};
+
+    EXPECT_THROW(image_vault.fetch_image(mp::FetchType::ImageOnly, query, stub_prepare, stub_monitor),
+                 std::runtime_error);
+}
+
+TEST_F(LXDImageVault, throws_with_non_alias_queries)
+{
+    ON_CALL(*mock_network_access_manager.get(), createRequest(_, _, _)).WillByDefault([](auto...) {
+        return new mpt::MockLocalSocketReply(mpt::not_found_data, QNetworkReply::ContentNotFoundError);
+    });
+
+    mp::Query query{"", "", false, "", mp::Query::Type::HttpDownload};
+
+    mp::LXDVMImageVault image_vault{hosts, mock_network_access_manager.get(), base_url};
+
+    EXPECT_THROW(image_vault.fetch_image(mp::FetchType::ImageOnly, query, stub_prepare, stub_monitor),
+                 std::runtime_error);
+}
+
+TEST_F(LXDImageVault, throws_with_invalid_remote)
+{
+    ON_CALL(*mock_network_access_manager.get(), createRequest(_, _, _)).WillByDefault([](auto...) {
+        return new mpt::MockLocalSocketReply(mpt::not_found_data, QNetworkReply::ContentNotFoundError);
+    });
+
+    mp::Query query{"", "foo", false, "bar", mp::Query::Type::Alias};
+
+    mp::LXDVMImageVault image_vault{hosts, mock_network_access_manager.get(), base_url};
+
+    EXPECT_THROW(image_vault.fetch_image(mp::FetchType::ImageOnly, query, stub_prepare, stub_monitor),
+                 std::runtime_error);
+}
+
+TEST_F(LXDImageVault, does_not_download_if_image_exists)
+{
+    ON_CALL(*mock_network_access_manager.get(), createRequest(_, _, _)).WillByDefault([](auto, auto request, auto) {
+        auto op = request.attribute(QNetworkRequest::CustomVerbAttribute).toString();
+        auto url = request.url().toString();
+
+        if (op == "GET")
+        {
+            if (url.contains("1.0/images/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"))
+            {
+                return new mpt::MockLocalSocketReply(mpt::image_info_data);
+            }
+        }
+        else if (op == "POST" && url.contains("1.0/images"))
+        {
+            // It should not try to request an image download
+            ADD_FAILURE();
```

Comment:
I believe this works:

```suggestion
            ADD_FAILURE() << "Image download shouldn't be requested";
```

Reply:
Sure, adding a helpful message in case this **does** fail is good.